### PR TITLE
[ADD] Module Sale Invoice Group Method

### PR DESCRIPTION
**Pull request by rgarnau**
Sale Invoice Group Method
=========================

This module allows you to combine several Sales Orders into a single invoice,
if they meet the split criteria defined by the 'Invoice Group Method'.

The split criteria is defined in the 'Invoice Group Method' by a combination
of fields of the Sales Order. 'Invoice Address' and 'Payment Term' are always
included.

You can assign a default 'Invoice Group Method' to customers, so that it will
be proposed on their orders.

When no Invoice Group Method is defined in a Sales Order, the standard
approach will be used, which groups by Sales Order.

Configuration
=============

#. Go to 'Sales / Configuration / Invoice Group Method'
#. Create an Invoice Group Method and choose the fields of the Sales Order
   that should be equal in that particular grouping method, for their orders
   to be merged into the same invoice. For example, create the Invoice Group
   Method 'By Customer' and select the field 'Customer').

Usage
=====

Update customers
----------------

#. Go to 'Sales / Sales / Customers'
#. Choose a customer and go to 'Sales & Purchases' page.
#. Update the 'Default Invoice Group Method'

Create a Quote / Sale Order
---------------------------

#. Go to 'Sales / Sales / Quotations'.
#. Create a new Quote and inside the 'Other Information' page select an
   option from the field 'Invoice Group Method'. If the customer had a
   default, it will have been provided already.
#. Complete the Quote as usual.

Create Invoices
---------------

#. Go to 'Sales / Sales / Sales Orders'.
#. Select all the sales orders with status 'To invoice'.
#. Go to 'Action' and click 'Invoice Order'. As a result, draft invoices will be
   created for the selected Sales Orders, consolidating them according to the
   Invoice Group Method defined.

Note: Existing draft invoices are not considered in the process of grouping.

@jbeficent @etobella

### CODE REVIEW

File: sale_invoice_group_method/README.rst [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+of fields of the Sales Order. 'Invoice Address', 'Currency' and 'Payment Term'
+are always included.
+
+You can assign a default 'Invoice Group Method' to customers, so that it will
+be proposed on their orders.
+
+When no Invoice Group Method is defined in a Sales Order, the standard
+approach will be used, which groups by Sales Order.
+
+Note: Existing draft invoices are not considered in the process of grouping.
```
Comment (JordiBForgeFlow) [inline sale_invoice_group_method/README.rst:21]: Maybe you can point out here to the other module that will consider draft invoices?

File: sale_invoice_group_method/README.rst [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+Configuration
+=============
+
+#. Go to 'Sales / Configuration / Invoice Group Method'
```
Comment (JordiBForgeFlow) [inline sale_invoice_group_method/README.rst:28]: In the actual menu it appears as 'Invoice groups'. Can you rename in the menu to what is advertised in the readme?

File: sale_invoice_group_method/model/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    _inherit = 'sale.order'
+
+    invoice_group_method_id = fields.Many2one(
+        string='Invoice Group Method',
+        comodel_name='sale.invoice.group.method'
+    )
+
+    @api.model
+    def create(self, vals):
+        res = super(SaleOrder, self).create(vals)
```
Comment (JordiBForgeFlow) [inline sale_invoice_group_method/model/sale_order.py]: Why do you propose the invoice group method here? The correct place would be on the onchange method associated to the partner: https://github.com/odoo/odoo/blob/10.0/addons/sale/models/sale.py#L196

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Technical and functional review. Minor comments.

**Comment by rgarnau** [comment]
@jbeficent changes applied! Thanks!

**Comment by JuaniFreedoo** [comment]
Great Jordi! this module does exactly what I need, from several orders, creates one invoice  per each order.
Very usefull indeed as well as flexible for other purposes. Thanks for your contribution.

**Comment by JordiBForgeFlow** [comment]
@JuaniAizean and you can combine it with https://github.com/OCA/sale-workflow/pull/551.

**Comment by rgarnau** [comment]
Hi @JuaniAizean, I see you have tried this module. Could you please leave your review? Thanks!

**Review by etobella — APPROVED** [review]
LGTM 👍

File: sale_invoice_group_method/view/res_partner_view.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        <field name="inherit_id" ref="base.view_partner_form"/>
+        <field name="arch" type="xml">
+            <xpath expr="//notebook/page[@name='sales_purchases']//group[@name='sale']//field[@name='message_bounce']"
+                   position="after">
+                <field name="invoice_group_method_id"/>
+            </xpath>
+        </field>
+    </record>
+
+</odoo>
```
Comment (MiquelRForgeFlow) [inline sale_invoice_group_method/view/res_partner_view.xml:20]: octicons_4-3-0_no-newline_24_0_fb1801_none

File: sale_invoice_group_method/view/sale_invoice_group_method_view.xml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    <record model="ir.actions.act_window" id="sale_invoice_group_method_action">
+        <field name="name">Sale Invoice Group Method</field>
+        <field name="type">ir.actions.act_window</field>
+        <field name="res_model">sale.invoice.group.method</field>
+        <field name="view_type">form</field>
+        <field name="view_mode">tree,form</field>
+        <field name="context">{'sale_invoice_group_method':1}</field>
+    </record>
+
+</odoo>
```
Comment (MiquelRForgeFlow) [inline sale_invoice_group_method/view/sale_invoice_group_method_view.xml:61]: octicons_4-3-0_no-newline_24_0_fb1801_none

**Review by MiquelRForgeFlow — APPROVED** [review]

**Pull request merged**